File: 01.patterns/pattern-21.js
// example pattern
// *****
// *   *
// *   *
// *   *
// *****

let n = 5
for(let i = 0; i < n; i++){
    let stars = ""
    for(let j = 0; j < n; j++){
        // if(i === 0 || i === n - 1){
        //     stars += "*"
        // }else{
        //     if(j === 0 || j === n - 1){
        //         stars += "*"
        //     }else{
        //         stars += " "
        //     }
        // }
        if(i === 0 || i === n - 1 || j === 0 || j === n - 1){
            stars += "*"
        }else{
            stars += " "
        }
    }
    console.log(stars)
}


// *****
// *   *
// *   *
// *   *
// *****

console.log("--- revision starts here ---")

let k = 5
for(let i = 0; i < k; i++){
    let row = ""
    for(let j = 0; j < k; j++){
        if(i === 0 || i === k - 1){
            row += "*"
        }else{
            if(j === 0 || j === k - 1){
                row += "*"
            }else{
                row += " "
            }
        }
    }
    console.log(row)
}


// final revision
// *****
// *   *
// *   *
// *   *
// *****

let rows = 5
for(let i = 0; i < rows; i++){
    let spaces = ''
    for(let j = 0; j < rows - 2; j++){
        spaces += '  '
    }
    let stars = ''
    if(i === 0 || i === rows - 1){
        for(let j = 0; j < rows; j++){
            stars += '* '
        }
        console.log(stars)
    }else{
        stars = '* '
        console.log(stars + spaces + stars)
    }
}

// practise for a job application
console.log("-----for job application-----")

// *****
// *   *
// *   *
// *   *
// *****

let p = 5

for(let i = 0; i < p; i++){
    let row = ''
    if(i === 0 || i === p - 1){
        for(let j = 0; j < p; j++){
            row += '*'
        }
    }else{
        for(let j = 0; j < p; j++){
            row += j === 0 || j === p - 1 ? `*` : ` `
        }
    }
    console.log(row)
}


// ------------------ revision with js and go -------------------------

console.log("------------------ revision with js and go -------------------------")


// *****
// *   *
// *   *
// *   *
// *****

let row = 5

for(let i = 0; i < row; i ++){
    let char = ''
    let space = ''
    for(let j = 0; j < row - 2; j++){
        space += ' '
    }
    if(i === 0 || i === row - 1){
        for(let j = 0; j < row; j++){
            char += '*'
        }
        console.log(char)
        continue
    }

    char += '*'
    console.log(char + space + char)
}

File: 01.patterns/pattern-21.go
// ------------------ revision with js and go -------------------------
package main

import (
	"fmt"
	"strings"
)

// *****
// *   *
// *   *
// *   *
// *****

func main() {
	p := 5

	// // for i := 0; i < p; i++ {

	// // }

	// for i := 0; i < p; i++ {
	// 	space := strings.Repeat("  ", p-i-1)
	// 	stars := strings.Repeat("* ", 2*i+1)
	// 	fmt.Println(space + stars)
	// }

	// for i := 0; i < 2*p-1; i++ {
	// 	var stars string
	// 	if i < p {
	// 		stars = strings.Repeat("* ", i+1)
	// 	} else {
	// 		stars = strings.Repeat("* ", 2*p-i-1)
	// 	}
	// 	fmt.Println(stars)
	// }

	// *****
	// *   *
	// *   *
	// *   *
	// *****

	fmt.Println("---------")
	for i := 0; i < p; i++ {
		if i == 0 || i == p-1 {
			result := strings.Repeat("*", p)
			fmt.Println(result)
			continue
		}
		space := strings.Repeat(" ", p-2)
		star := "*"
		fmt.Println(star + space + star)
	}

}
